src/main/java/org/jenkinsci/plugins/credentialsbinding/masking/PatternMaskerProvider.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2019 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.credentialsbinding.masking;
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.TreeSet;
+import java.util.regex.Pattern;
+
+/**
+ * Quotes strings according to quotation rules of various programs. Sometimes confused with escaping, quoting is done
+ * to pass literal strings in a shell or other interpreter.
+ *
+ * @since TODO
+ */
+public interface PatternMaskerProvider extends ExtensionPoint {
+
+    @Nonnull Collection<String> getAlternativeForms(@Nonnull String input);
+
+    static @Nonnull ExtensionList<PatternMaskerProvider> all() {
+        return ExtensionList.lookup(PatternMaskerProvider.class);
+    }
+
+    @Restricted(NoExternalUse.class)
+    static @Nonnull Collection<String> getAllAlternateForms(@Nonnull String input) {
+        Collection<String> alternateForms = new TreeSet<>(BY_REVERSE_LENGTH);
+        for (PatternMaskerProvider provider : all()) {
+            alternateForms.addAll(provider.getAlternativeForms(input));
+        }
+        return alternateForms;
+    }
+
+    @Restricted(NoExternalUse.class)
+    static @Nonnull Pattern getMaskingPattern(@Nonnull Collection<String> inputs) {
+        Collection<String> patterns = new TreeSet<>(BY_REVERSE_LENGTH);
```

Comment:
You should make explicit the requirement to have this ordered collection (variable name, comment?). 
Two points to mention:

1. If you have two secrets with one being a substring of the other, there is no guarantee the longest will be masked completely (should be part of the documentation around limitation)
2. The reverse length order is used in case you have for a given secret multiple patterns of different length, like `abc` and `'abc'`, you do not want to have the shorter version used before the longer and thus, giving information about the password content in the log.

The second point is particularly interesting as it's not demonstrated in the test, but it should be from my PoV.

Reply:
My main use case for this ordering is so that `abc` and `'abc'` are not masked as `****` and `'****'` respectively. I try to include the surrounding quotes, though now I wonder if that's leaking information or not.

Reply:
Oh I see what you mean. I'll updated the javadoc here.